### PR TITLE
Port shared-memory apps (bfs, cc, jaccard, kcore, pr, sssp) to use PropertyGraphs

### PR DESCRIPTION
**Pull request by gurbinder533**

### CODE REVIEW

File: lonestar/analytics/cpu/bfs/bfs.cpp [hunk 1]
```diff
@@ -142,17 +137,15 @@ struct OneTilePushWrap {
...
+    EdgeTile t{graph->edge_begin(n), graph->edge_end(n)};
 
     cont.push(t);
   }
 };
 
 template <bool CONCURRENT, typename T, typename P, typename R>
 void
-asyncAlgo(Graph& graph, GNode source, const P& pushWrap, const R& edgeRange) {
+asyncAlgo(Graph* graph, GNode source, const P& pushWrap, const R& edgeRange) {
```
Comment (ddn0) [inline lonestar/analytics/cpu/bfs/bfs.cpp:148]: FWIW, I don't know if it is totally worth it to change all uses of `T&` to `T*` for old code. It is pretty noisy, but I'll let you decide if it is worth it.

File: libgalois/include/galois/graphs/PropertyGraph.h [hunk 2]
```diff
@@ -49,6 +49,7 @@ class PropertyGraph {
   using edge_iterator = boost::counting_iterator<uint64_t>;
   using edges_iterator = StandardRange<NoDerefIterator<edge_iterator>>;
   using iterator = node_iterator;
+  using GraphNode = uint32_t;
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/PropertyGraph.h:52]: just "Node" or "Vertex" is better in my opinion

File: lonestar/analytics/cpu/bfs/bfs.cpp [hunk 3]
```diff
@@ -334,21 +328,20 @@ main(int argc, char** argv) {
...
-  size_t approxNodeData = 4 * (graph.size() + graph.sizeEdges());
+  size_t approxNodeData = 4 * (graph.num_nodes() + graph.num_edges());
   galois::preAlloc(
       8 * numThreads + approxNodeData / galois::runtime::pagePoolSize());
 
   galois::reportPageAlloc("MeminfoPre");
 
-  galois::do_all(galois::iterate(graph), [&graph](GNode n) {
-    graph.getData(n) = BFS::DIST_INFINITY;
+  galois::do_all(galois::iterate(graph.begin(), graph.end()), [&graph](auto n) {
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/bfs/bfs.cpp:342]: Not sure whether this is still active/working but `iterate(graph)` semantics are different from `iterate(graph.begin(), graph.end())` : former uses local ranges, latter does not. Local ranges are (by default) balanced based on edges, so if edges are NOT being iterated within an operator, then typically explicit begin() and end() without stealing is better for performance.

File: lonestar/liblonestar/include/Lonestar/Utils.h [hunk 4]
```diff
@@ -36,23 +43,76 @@ const uint32_t SourcePicker<Graph>::kRandSeed = 27491095;
...
+
+  if (auto r = pfg->AddNodeProperties(node_table); !r) {
+    return r.error();
+  }
+  return galois::ResultSuccess();
+}
+
+template <typename NodeProps>
+std::unique_ptr<galois::graphs::PropertyFileGraph>
+MakeFileGraph(
```
Comment (roshandathathri) [inline lonestar/liblonestar/include/Lonestar/Utils.h:101]: It's better to have MakePropertyGraph() and return in-memory PropertyGraph directly instead of PropertyFileGraph.

**Review by roshandathathri — APPROVED** [review]
LGTM

File: lonestar/analytics/cpu/bfs/CMakeLists.txt [hunk 5]
```diff
@@ -2,12 +2,12 @@ add_executable(bfs-cpu bfs.cpp)
...
+add_test_scale(small1 bfs-cpu "${BASEINPUT}/propertygraphs/rmat15")
+#add_test_scale(small2 bfs-cpu "${BASEINPUT}/propertygraphs/rmat15")
 
-add_executable(bfs-directionopt-cpu bfsDirectionOpt.cpp)
-add_dependencies(apps bfs-directionopt-cpu)
-target_link_libraries(bfs-directionopt-cpu PRIVATE Galois::shmem lonestar)
-install(TARGETS bfs-directionopt-cpu DESTINATION "${CMAKE_INSTALL_BINDIR}" COMPONENT apps EXCLUDE_FROM_ALL)
-add_test_scale(small1 bfs-directionopt-cpu "${BASEINPUT}/reference/structured/rome99.gr")
-add_test_scale(small2 bfs-directionopt-cpu "${BASEINPUT}/scalefree/rmat10.gr")
+#add_executable(bfs-directionopt-cpu bfsDirectionOpt.cpp)
```
Comment (ddn0) [inline lonestar/analytics/cpu/bfs/CMakeLists.txt:8]: what do we need to do for this to compile?

File: lonestar/liblonestar/include/Lonestar/Utils.h [hunk 6]
```diff
@@ -36,23 +43,76 @@ const uint32_t SourcePicker<Graph>::kRandSeed = 27491095;
 //! (https://github.com/sbeamer/gapbs/blob/master/src/tc.cc WorthRelabelling())
 template <typename Graph>
 bool
-isApproximateDegreeDistributionPowerLaw(const Graph& graph) {
-  uint32_t averageDegree = graph.sizeEdges() / graph.size();
+isApproximateDegreeDistributionPowerLaw(Graph* graph) {
```
Comment (ddn0) [inline lonestar/liblonestar/include/Lonestar/Utils.h:46]: why isn't this `const Graph&` any more?

File: lonestar/liblonestar/include/Lonestar/Utils.h [hunk 7]
```diff
@@ -36,23 +43,76 @@ const uint32_t SourcePicker<Graph>::kRandSeed = 27491095;
...
   double sample_median = samples[num_samples / 2];
   return sample_average / 1.3 > sample_median;
 }
+
+/**
+   * Constructs the node properties based on tuple on U.
+   */
+template <typename U>
+galois::Result<void>
+constructNodeProperties(
```
Comment (ddn0) [inline lonestar/liblonestar/include/Lonestar/Utils.h:73]: `ConstructNodeProperties`?

File: lonestar/liblonestar/include/Lonestar/Utils.h [hunk 8]
```diff
@@ -5,6 +5,13 @@
 #include <random>
 #include <vector>
 
+#include <arrow/stl.h>
```
Comment (ddn0) [inline lonestar/liblonestar/include/Lonestar/Utils.h:8]: FWIW, `Utils.h` is a pretty generic name and a path for cyclic dependencies as there is no clear theme for the contents of a utility header.

Some of the constructor-like methods might fit better with as static methods on the class instances they are constructing or at least in the corresponding header file.

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 2 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyGraph.h:52]: "Node" is used everywhere actually.

File: lonestar/analytics/cpu/bfs/bfs.cpp (hunk 3 above)
Comment (gurbinder533) [reply lonestar/analytics/cpu/bfs/bfs.cpp:342]: I don't think local ranges are being set at the moment. I was going to use iterate(graph) as it looks much cleaner.

File: lonestar/analytics/cpu/bfs/CMakeLists.txt (hunk 5 above)
Comment (gurbinder533) [reply lonestar/analytics/cpu/bfs/CMakeLists.txt:8]: We need both in-edges and out-edges for direction optimization. It is using LC_CSR_CSC_Graph.

File: lonestar/liblonestar/include/Lonestar/Utils.h (hunk 4 above)
Comment (gurbinder533) [reply lonestar/liblonestar/include/Lonestar/Utils.h:101]: I think it is better to return the PropertryFileGraph, add the required node/edge properties, shuffle/sort topology, and then put a view on top of it.

File: lonestar/liblonestar/include/Lonestar/Utils.h (hunk 8 above)
Comment (gurbinder533) [reply lonestar/liblonestar/include/Lonestar/Utils.h:8]: Okay. Also, I am thinking of adding member functions `ConstructNodeProperties` and `ConstructEdgeProperties` to the PropertyFileGraph class itself. Any objections?

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 2 above)
Comment (roshandathathri) [reply libgalois/include/galois/graphs/PropertyGraph.h:52]: It's better to change now than later. But let's decide what we want. What do you prefer?

@ddn0 what do you prefer?

File: lonestar/analytics/cpu/bfs/bfs.cpp (hunk 3 above)
Comment (roshandathathri) [reply lonestar/analytics/cpu/bfs/bfs.cpp:342]: Ok

File: lonestar/liblonestar/include/Lonestar/Utils.h (hunk 4 above)
Comment (roshandathathri) [reply lonestar/liblonestar/include/Lonestar/Utils.h:101]: I'm slightly confused by it. Node or edge properties can be added to PropertyGraph as well, right?

Comment (gurbinder533) [reply lonestar/liblonestar/include/Lonestar/Utils.h:101]: No, PropertyGraph provides a view on the fully constructed PropertyFileGraph with all the node and edge properties. @ddn0 Please correct me if I am wrong.

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 2 above)
Comment (ddn0) [reply libgalois/include/galois/graphs/PropertyGraph.h:52]: "Node" would be better, because "Graph" is clear from context since "Node" a field on a "Graph" class. I think node is used more frequently than vertex in the codebase, so that's an argument for node over vertex.

The counterargument to changing GraphNode at all would be that this would affect a lot of files, but I do agree with @roshandathathri here that if we want more consistency in our API now is the time to do it.

File: lonestar/liblonestar/include/Lonestar/Utils.h (hunk 4 above)
Comment (ddn0) [reply lonestar/liblonestar/include/Lonestar/Utils.h:88]: Yes, the current design is that a PropertyGraph is just a view on top of a PropertyFileGraph. If you want to add, remove or filter properties, it should be done at the PropertyFileGraph level rather than at PropertyGraph.

Comment (roshandathathri) [reply lonestar/liblonestar/include/Lonestar/Utils.h:88]: Ok, makes sense.

File: libgalois/include/galois/graphs/PropertyGraph.h [hunk 9]
```diff
@@ -47,8 +47,10 @@ class PropertyGraph {
   using edge_properties = EdgeProps;
   using node_iterator = boost::counting_iterator<uint32_t>;
   using edge_iterator = boost::counting_iterator<uint64_t>;
+  using const_edge_iterator = std::add_cv<edge_iterator>::type;
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/PropertyGraph.h:50]: Why is volatile required?

File: lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp [hunk 10]
```diff
@@ -109,19 +109,19 @@ main(int argc, char** argv) {
...
-    autoAlgoTimer.start();
-    if (isApproximateDegreeDistributionPowerLaw(degreeGraph)) {
-      algo = Async;
-    } else {
-      algo = Level;
-    }
-    autoAlgoTimer.stop();
-    galois::gInfo("Choosing ", ALGO_NAMES[algo], " algorithm");
-  }
+  // if (algo == AutoAlgo) {
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp:112]: Why?

File: lonestar/analytics/cpu/triangle-counting/Triangles.cpp [hunk 11]
```diff
@@ -467,14 +467,14 @@ makeSortedGraph(Graph& graph) {
...
   galois::StatTimer autoAlgoTimer("AutoAlgo_0");
-  if (!relabel) {
-    galois::graphs::FileGraph degreeGraph;
-    degreeGraph.fromFile(inputFile);
-    degreeGraph.initNodeDegrees();
-    autoAlgoTimer.start();
-    relabel = isApproximateDegreeDistributionPowerLaw(degreeGraph);
-    autoAlgoTimer.stop();
-  }
+  // if (!relabel) {
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/triangle-counting/Triangles.cpp:470]: Why?

File: lonestar/liblonestar/include/Lonestar/Utils.h [hunk 12]
```diff
@@ -23,7 +49,7 @@ class SourcePicker {
     typename Graph::GraphNode source;
     do {
       source = udist(rng);
-    } while (graph.getDegree(source));
+    } while (std::distance(graph.edge_begin(source), graph.edge_end(source)));
```
Comment (roshandathathri) [inline lonestar/liblonestar/include/Lonestar/Utils.h:52]: Please add getDegree() as a member function

File: lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp (hunk 10 above)
Comment (gurbinder533) [reply lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp:112]: This requires changes in old FileGraph (due to isApproximateDegreeDistributionPowerLaw) which will not be used once we port them. I will enable it once we have ported it over to use PropertyGraph.

File: lonestar/analytics/cpu/triangle-counting/Triangles.cpp (hunk 11 above)
Comment (gurbinder533) [reply lonestar/analytics/cpu/triangle-counting/Triangles.cpp:470]: Same reason.

File: lonestar/liblonestar/include/Lonestar/Utils.h (hunk 12 above)
Comment (gurbinder533) [reply lonestar/liblonestar/include/Lonestar/Utils.h:52]: Currently we are only need getDegree() in isApproximateDegreeDistributionPowerLaw. Not sure if it is worth adding a new member function for that. What do you think?

File: lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp (hunk 10 above)
Comment (roshandathathri) [reply lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp:112]: Ok. Please add a comment in the commit message saying why you commented out certain code.

**Comment by gurbinder533** [comment]
I want to merge this PR before I move on to other algorithms as most of the remaining algorithms need topology shuffling/sorting operations which I am planning to work on next. Any objections?

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 2 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyGraph.h:53]: I will change it to "Node".

**Comment by roshandathathri** [comment]
> Any objections?

No, but please keep a note of the pending action items.

**Comment by ddn0** [comment]
If you are not planning to squash this PR, can you edit the commit named "WIP: Port shared-memory apps to use PropertyGraphs" to something that doesn't have the "WIP" in it?

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 9 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyGraph.h:50]: std::add_const is required. Fixed it.

**Comment by gurbinder533** [comment]
> If you are not planning to squash this PR, can you edit the commit named "WIP: Port shared-memory apps to use PropertyGraphs" to something that doesn't have the "WIP" in it?

I am planning to squash. Doing it now.

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 9 above)
Comment (ddn0) [reply libgalois/include/galois/graphs/PropertyGraph.h:50]: I'm actually wondering why this is needed at all. Iterators can be passed by value so `const`-ness isn't particular important for them unless you mean an iterator that when dereferrenced returns `const T&`, in which case, `std::add_cv`/`std::add_const` isn't going to do what you want anyway.

File: libgalois/include/galois/graphs/PropertyFileGraph.h [hunk 13]
```diff
@@ -200,6 +201,13 @@ class GALOIS_EXPORT PropertyFileGraph {
   Result<void> AddNodeProperties(const std::shared_ptr<arrow::Table>& table);
   Result<void> AddEdgeProperties(const std::shared_ptr<arrow::Table>& table);
 
+  template <typename NodeProps>
+  Result<void> ConstructNodeProperties(
```
Comment (ddn0) [inline libgalois/include/galois/graphs/PropertyFileGraph.h:205]: I don't know how I feel about this interface. PropertyFileGraph is mostly in terms of libarrow concepts so using a properties here seems out of place. What's wrong with using `AddNodeProperties`/`AddEdgeProperties` here?

File: libgalois/include/galois/graphs/PropertyGraph.h [hunk 14]
```diff
@@ -84,14 +91,42 @@ class PropertyGraph {
...
+   *
+   * @param node node to get the edge of
+   * @returns iterator to the end of the edges of node, i.e. the first edge of
+   *     the next node (or an "end" iterator if there is no next node)
+   */
+  edge_iterator edge_end(Node node) {
+    return pfg_->topology().edge_range(node).second;
+  }
+
+  const_edge_iterator edge_end(Node node) const {
```
Comment (ddn0) [inline libgalois/include/galois/graphs/PropertyGraph.h:119]: if `const_edge_iterator` is just `const edge_iterator` I don't see the point of having two methods here rather than just:
```
edge_iterator edge_end(....) const
```

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 9 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyGraph.h:50]: So, this is required when passing const Graph& to a function (such as "isApproximateDegreeDistributionPowerLaw") and calling edge_begin/edge_end which returns T& and the compiler doesn't like it. Is there a better way to do this?

File: libgalois/include/galois/graphs/PropertyGraph.h [hunk 15]
```diff
@@ -84,14 +91,42 @@ class PropertyGraph {
...
+  }
+
+  /**
+   * Gets the end edge boundary of some node.
+   *
+   * @param node node to get the edge of
+   * @returns iterator to the end of the edges of node, i.e. the first edge of
+   *     the next node (or an "end" iterator if there is no next node)
+   */
+  edge_iterator edge_end(Node node) {
```
Comment (ddn0) [inline libgalois/include/galois/graphs/PropertyGraph.h:115]: do we really need these functions if we already have `edges()`? At the very least can we just write this in terms of `edges()`?

File: libgalois/include/galois/graphs/PropertyFileGraph.h (hunk 13 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyFileGraph.h:205]: I don't understand. I do need to construct arrow::table before calling AddNodeProperties/AddEdgeProperties, isn't it?

Comment (ddn0) [reply libgalois/include/galois/graphs/PropertyFileGraph.h:205]: Yes. We simply need a function that creates a table of a specific arrow type and length, yet this function hides this fact behind the template parameter `NodeProps` and makes it harder for users to initialize their data if they wanted to.

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 14 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyGraph.h]: Fixed. Thank you.

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 15 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyGraph.h]: Yes, some of the applications are using `edge_begin` and `edge_end` and I feel they are more intuitive in some cases. But I have re-written them to use `edges`.

File: libgalois/include/galois/graphs/PropertyFileGraph.h (hunk 13 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyFileGraph.h:205]: Makes sense. I started with something similar to this function. Let me do that. I will keep it in Properties.h.

File: libgalois/include/galois/graphs/PropertyGraph.h (hunk 9 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyGraph.h]: Don't need this anymore.

File: libgalois/include/galois/graphs/PropertyFileGraph.h (hunk 13 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyFileGraph.h]: @ddn0: Done. Please make a pass. Thank you.

File: libgalois/include/galois/Properties.h [hunk 16]
```diff
@@ -355,5 +356,29 @@ struct LargeStringReadOnlyProperty {
   using ViewType = StringPropertyReadOnlyView<int64_t>;
 };
 
+template <typename Props>
+Result<std::shared_ptr<arrow::Table>>
+AllocateTable(
+    uint64_t num_nodes,
```
Comment (ddn0) [inline libgalois/include/galois/Properties.h:362]: `num_rows` probably? You can use this to allocate edge or node properties.

File: libgalois/include/galois/Properties.h [hunk 17]
```diff
@@ -355,5 +356,29 @@ struct LargeStringReadOnlyProperty {
...
+    std::enable_if_t<std::tuple_size_v<Props> != 0>* = nullptr) {
```
Comment (ddn0) [inline libgalois/include/galois/Properties.h:363]: I'd remove this `enable_if` part. 

If you still want static checking, a `static_assert(std::tuple_size_v<Props> != 0)` is just as good and allows you to have a better compiler error message. Otherwise, you can also just return a runtime error if `num_tuple_elem` is 0. The latter approach doesn't give you compile-time checking (unless `arrow::stl::TableFromTuple` itself has an issue?) but users already have to check for errors anyway.

File: libgalois/include/galois/Properties.h [hunk 18]
```diff
@@ -355,5 +356,29 @@ struct LargeStringReadOnlyProperty {
...
+  std::shared_ptr<arrow::Table> table;
+  // Construct node property data
```
Comment (ddn0) [inline libgalois/include/galois/Properties.h:365]: this comment is not needed or it you think it is, we should first start with documenting the function interface before going into documenting the implementation details.

File: libgalois/include/galois/graphs/PropertyFileGraph.h [hunk 19]
```diff
@@ -6,6 +6,7 @@
 #include <vector>
 
 #include <arrow/api.h>
+#include <arrow/stl.h>
```
Comment (ddn0) [inline libgalois/include/galois/graphs/PropertyFileGraph.h:9]: why is this needed here rather than in a file that explicitly uses something from the `arrow::stl` namespace?

File: lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp [hunk 20]
```diff
@@ -28,7 +28,7 @@ constexpr static const char* const REGION_NAME = "BC";
 
 enum Algo { Level = 0, Async, Outer, AutoAlgo };
 
-const char* const ALGO_NAMES[] = {"Level", "Async", "Outer", "Auto"};
+// const char* const ALGO_NAMES[] = {"Level", "Async", "Outer", "Auto"};
```
Comment (ddn0) [inline lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp:31]: is this used at all? maybe delete.

File: lonestar/analytics/cpu/connected-components/ConnectedComponents.cpp [hunk 21]
```diff
@@ -142,81 +142,90 @@ struct Node : public galois::UnionFindNode<Node> {
     return *this;
   }
 
-  component_type component() { return this->get(); }
+  ComponentType component() { return this->get(); }
   bool isRepComp(unsigned int) { return false; }
 };
 
-const unsigned int LABEL_INF = std::numeric_limits<unsigned int>::max();
+[[maybe_unused]] const unsigned int LABEL_INF =
```
Comment (ddn0) [inline lonestar/analytics/cpu/connected-components/ConnectedComponents.cpp:149]: why `[[maybe_unused]]`?

File: lonestar/analytics/cpu/connected-components/ConnectedComponents.cpp [hunk 22]
```diff
@@ -508,23 +544,32 @@ struct BlockedAsyncAlgo {
...
-    galois::graphs::readGraph(graph, inputFile);
+  using NodeData = std::tuple<NodeComponent>;
+  using EdgeData = std::tuple<>;
+  typedef galois::graphs::PropertyGraph<NodeData, EdgeData> Graph;
+  typedef typename Graph::Node GNode;
+  using Edge = std::pair<GNode, typename Graph::edge_iterator>;
+
+  void Initialize(Graph* graph) {
+    galois::do_all(galois::iterate(*graph), [&](const GNode& node) {
+      graph->GetData<NodeComponent>(node) = new Node();
```
Comment (ddn0) [inline lonestar/analytics/cpu/connected-components/ConnectedComponents.cpp:571]: do we `delete` these pointers at any point?

File: libgalois/include/galois/graphs/PropertyFileGraph.h (hunk 19 above)
Comment (gurbinder533) [reply libgalois/include/galois/graphs/PropertyFileGraph.h:9]: Yes, it is not needed here anymore. I moved it to Properties.h.

File: lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp (hunk 20 above)
Comment (gurbinder533) [reply lonestar/analytics/cpu/betweennesscentrality/BetweennessCentrality.cpp:31]: This is temporarily commented out. I will need it once I port this application to use property graphs. I will leave a comment there.

File: lonestar/analytics/cpu/connected-components/ConnectedComponents.cpp (hunk 21 above)
Comment (gurbinder533) [reply lonestar/analytics/cpu/connected-components/ConnectedComponents.cpp:149]: I was debugging something. Good catch. Thanks.

File: lonestar/analytics/cpu/connected-components/ConnectedComponents.cpp (hunk 22 above)
Comment (gurbinder533) [reply lonestar/analytics/cpu/connected-components/ConnectedComponents.cpp:602]: Good catch.
